tests/python/test_python_frontend.py

```diff
@@ -4458,6 +4457,86 @@ def fusion_func(fd: FusionDefinition):
         for i in range(num_out):
             self.assertEqual(nvf_out[i].data_ptr(), inputs[0].data_ptr())
 
+    @unittest.skipIf(is_pre_ampere(), "Only supported on Ampere and newer devices.")
+    def test_sdpa_fwd(self):
+        def fusion_func(
+            fd: FusionDefinition, has_dropout: bool, has_causal: bool, has_scale: bool
+        ):
+            q = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            k = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            v = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            dropout_p, is_causal, scale = None, None, None
+            if has_dropout:
+                dropout_p = fd.define_scalar(None, dtype=DataType.Double)
+            if has_causal:
+                is_causal = fd.define_scalar(None, dtype=DataType.Bool)
+            if has_scale:
+                scale = fd.define_scalar(None, dtype=DataType.Double)
+            attn, *intermediate_results = fd.ops.sdpfa_fwd(
+                q, k, v, dropout_p, is_causal, scale
+            )
+            fd.add_output(attn)
+
+        N, H, L, S, E = 4, 8, 16, 16, 8
+        qkv = [
+            torch.randn((N, H, L, E), dtype=torch.bfloat16, device="cuda:0"),
+            torch.randn((N, H, S, E), dtype=torch.bfloat16, device="cuda:0"),
+            torch.randn((N, H, S, E), dtype=torch.bfloat16, device="cuda:0"),
+        ]
+
+        dropout_vals = [None, 0.0, 0.2]
+        is_causal_vals = [None, True, False]
+        scale_vals = [None, 1 / E**0.5, 1e-3]
+        torch.manual_seed(0)
```

Comment:
Is this necessary and why?

Reply:
1. Setting torch.manual_seed(0) since it is currently set within `exec_nvfuser`.
Will refactor that to set for the whole class in PR #2701. A constant seed is required for testing with non-zero dropout.

2. Values for dropout, is_causal, scale are to exercise different combinations of these parameters and verify correctness

Reply:
> torch.manual_seed(0) is currently set within exec_nvfuser.

I missed that fact -- thanks for clarifying. However, I'm even more confused on how we are using RNG in the test. 
1. This particular `torch.manual_seed(0)` is done before the first subtest. But `exec_nvfuser` is executed before each subtest. There seems to be an asymmetry and I don't understand how they are logically connected. 
2. `fork_rng` should surround either actual or ref, but not both (at least not necessary). It essentially snapshots the state at entry and restores that state at exit. This restore is only needed between actual and ref, but not after them.

Reply:
I'm unsure about this, but I now start to think it may be too early to use fork_rng before we remove the manual_seed(0) from exec_nvfuser. I suspect how we use RNG today is broken based on the previous comment. Although fork_rng is the right solution, adding that on top of the brokenness **at this very moment** may create more troubles.

Reply:
(happy to discuss live)

Reply:
I *think* the issue here is that we are running two references....

i.e. `exec_nvfuser` runs a reference for serde, while this test runs another reference with torch.

We should probably be using two `fork_rng` inside exec_nvfuser for nvfuser run and the serde reference. That'll render the manual seed here obselete and we can also drop both fork_rng in this test

More than happy to have that in a separate PR. :stuck_out_tongue_winking_eye:

Reply:
Got it. 

Sorry that I opened the can of worm. I'm now convinced that we'd better attempt fork_rng in a separate PR and only use manual_seed in this PR. The required change for fork_rng is larger than I expected, and a mix of fork_rng and manual_seed as is is confusing.

Reply:
Notes to myself: currently, we run five implementations of a fusion and expect them to match: 
```
manual_seed(0)
run fusion without serde
manual_seed(0)
run fusion with serde
manual_seed(0)
run captured fusion without serde
manual_seed(0)
run captured fusion with serde
manual_seed(0)
run reference implementation 
```